### PR TITLE
Feature/tcan bridge

### PR DESCRIPTION
**Pull request by simonkerscher**
This creates nodes that convert CAN-messages to ROS2-messages and vice versa. Could be useful to connect CAN-controlled visualizers and input devices at remote locations. 

See this terminal output for the node in action. The clocks between computers are not synchronized. The delay introduced by the bridge and the networking (local Ethernet switch in this case) seems to be <5ms.

image

### CODE REVIEW

File: tcan_bridge/.clang-format [hunk 1]
```diff
@@ -0,0 +1 @@
+/home/kersimon/workspaces/.clang-format
```
Comment (djud) [inline tcan_bridge/.clang-format]: remove this file?

File: tcan_bridge/package.xml [hunk 2]
```diff
@@ -0,0 +1,18 @@
+<?xml version="1.0"?>
+<package format="3">
+  <name>tcan_bridge</name>
+  <version>0.0.0</version>
+  <description>TODO: Package description</description>
+  <maintainer email="simon.kerscher@mavt.ethz.ch">Simon Kerscher</maintainer>
+  <license>TODO: License declaration</license>
```
Comment (djud) [inline tcan_bridge/package.xml]: MIT as everything else?

File: tcan_bridge/package.xml [hunk 3]
```diff
@@ -0,0 +1,18 @@
+<?xml version="1.0"?>
+<package format="3">
+  <name>tcan_bridge</name>
+  <version>0.0.0</version>
+  <description>TODO: Package description</description>
```
Comment (djud) [inline tcan_bridge/package.xml]: fill in

File: tcan_bridge_msgs/package.xml [hunk 4]
```diff
@@ -0,0 +1,23 @@
+<?xml version="1.0"?>
+<?xml-model href="http://download.ros.org/schema/package_format3.xsd" schematypens="http://www.w3.org/2001/XMLSchema"?>
+<package format="3">
+  <name>tcan_bridge_msgs</name>
+  <version>0.0.0</version>
+  <description>TODO: Package description</description>
```
Comment (djud) [inline tcan_bridge_msgs/package.xml]: fill in description and license

File: tcan_bridge/src/bidirectional_bridge.cpp [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+#include <tcan_can/SocketBusOptions.hpp>
+
+#include <tcan_bridge/tcan_bridge.hpp>
+
+namespace tcan_bridge {
+
+class BidirectionalBridge : public rclcpp::Node, public TcanBridge {
+ public:
+  BidirectionalBridge() = delete;
+  BidirectionalBridge(const std::string& nodeName, const std::string& canInterfaceName, const std::string& publishedName,
```
Comment (djud) [inline tcan_bridge/src/bidirectional_bridge.cpp]: change `publishedName` to `publishedTopicName`

File: tcan_bridge/src/bidirectional_bridge.cpp [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+                      const std::string& subscribedName)
```
Comment (djud) [inline tcan_bridge/src/bidirectional_bridge.cpp]: change `subscribedName` to `subscribedTopicName`

File: tcan_bridge/src/bidirectional_bridge.cpp [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+  }
+
+  void rosMsgCallback(const tcan_bridge_msgs::msg::CanFrame::SharedPtr msg) {
+    tcan_can::CanMsg canMsg(msg->id, msg->length, msg->data.data());
+    canManager_.getCanBus(0)->sendMessage(canMsg);
+  }
+
+ private:
+  rclcpp::Publisher<tcan_bridge_msgs::msg::CanFrame>::SharedPtr publisher_;
+  std::vector<rclcpp::Subscription<tcan_bridge_msgs::msg::CanFrame>::SharedPtr> subscribers_;
```
Comment (djud) [inline tcan_bridge/src/bidirectional_bridge.cpp]: why is there a vector of subscribers if we only have one?

File: tcan_bridge/src/bidirectional_bridge.cpp [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+  if (argc != 4) {
+    std::cout << "Call with\n";
+    std::cout << "`ros2 run tcan_bridge bidirectional_bridge <canInterfaceName> <publishedRosTopicName> <subscribedRosTopicName>`\n";
+    std::cout << "e.g. `ros2 run tcan_bridge bidirectional_bridge can0 can0AsRos toPublishOnCan0`" << std::endl;
+    return -1;
+  }
+  const std::string canInterfaceName = argv[1];
+  const std::string publishedName = argv[2];
+  const std::string subscribedName = argv[3];
+  const std::string nodeName = canInterfaceName + "_to_" + publishedName + "_and_from_" + subscribedName + "_bridge_";
```
Comment (djud) [inline tcan_bridge/src/bidirectional_bridge.cpp]: interface, published and subscribed topic name should be ros2 parameters. They can then still be set through the commandline but we can also do proper config files

**Review by djud — CHANGES_REQUESTED** [review]
There is some duplication in the three executables (bidirectional, can to ros and ros to can). Imo this could be reduced to just the bidirectional bridge by adding a check to the publisher if there is a subscriber to the topic it publishes. The other direction does not have to be checked, becuase there will be no callback triggered if there are no messages published on the topics.

File: tcan_bridge/src/bidirectional_bridge.cpp (hunk 7 above)
Comment (simonkerscher) [reply tcan_bridge/src/bidirectional_bridge.cpp]: My plan was to have config files that allowed multiple forwardings with one node, however never finished this

File: tcan_bridge/src/bidirectional_bridge.cpp (hunk 8 above)
Comment (simonkerscher) [reply tcan_bridge/src/bidirectional_bridge.cpp]: Will do this tomorrow

**Review by djud — APPROVED** [review]

**Pull request merged**